week1/tictactoe/borboss366/Main.hs

```diff
@@ -0,0 +1,354 @@
+  outputUpperRow w
+  mapM_ (\fr -> outputFieldRow fr >> outputMiddleRow w) (init rows)
+  outputFieldRow (last rows)
+  outputLowerRow w
+
+  where
+    (w, _) = (getDimension.calcDimension) pos
+    rows = map (toCellLine pos) $ getRows (calcDimension pos)
+{-- End of visualization stuff --}
+
+{-- Parsing stuff --}
+parseX :: String -> Maybe Int
+parseX str  | length str == 1 && isAsciiLower c = Just $ ord c - ord 'a'
+            | otherwise = Nothing
+            where c = toLower $ head str
+
+
+parsePosition :: String -> Maybe TTTPoint
+parsePosition line
+  | length values == 2 =
+    do
+      let [xStr, yStr] = values
+      xm <- parseX xStr
+      ym <- readMaybe yStr
+      return (TTTPoint (xm, ym - 1))
+  | otherwise = Nothing
+  where
+    g c1 c2 = isDigit c1 == isDigit c2
+    values = groupBy g line
+
+-- Sorry, Artoym, I didn't understand how you made the following code work
+-- the binding [xStr, yStr] <- groupBy g line
+-- seems strange to me
+{--
+parsePosition line = do
+  let g c1 c2 = isDigit c1 == isDigit c2
+  [xStr, yStr] <- groupBy g line
+  xm <- parseX xStr
+  ym <- readMaybe yStr
+  return (TTTPosition x (ym-1))
+--}
+
+parseConfig :: String -> Maybe (Int, Int, Int)
+parseConfig line |  length values == 3 = do
+                      let [ws, hs, rs] = values
+                      w <- readMaybe ws
+                      h <- readMaybe hs
+                      r <- readMaybe rs
+                      return (w,h,r)
+                 |  otherwise = Nothing
+                 where values = words line
+
+readUntilGood :: String -> (String -> Maybe a) -> IO a
+readUntilGood title parseFun = do
+  putStrLn title
+  line <- getLine
+  case parseFun line of
+    Just object -> return object
+    Nothing -> readUntilGood title parseFun
+{-- End of parsing stuff--}
+
+{-- the solving functions --}
+
+solveFirstEmpty :: TTTPosition -> Maybe TTTPosition
+solveFirstEmpty pos =
+  case getPossibleMoves pos of
+    (m:_) -> Just $ makeMove pos m
+    [] -> Nothing
+
+solveMinimax :: TTTPosition -> Maybe TTTPosition
+solveMinimax pos =
+  case calcMoveState pos of
+    Move _ -> Just $ makeMove pos (fst bestMove)
+    _ -> Nothing
+  where
+    mvz = map (\m -> (m, negate $ minimax $ makeMove pos m)) (getPossibleMoves pos)
+    bestMove = maximumBy (compare `on` snd) mvz
+
+minimax :: TTTPosition -> Double
+minimax pos =
+  case calcMoveState pos of
+    Move _ -> snd ourMove
+    Win _ -> -1 -- opponent already won
+    TTTie -> 0
+  where
+    -- get the positive values
+    -- opp always chooses the move with maximum
+    makeMoveCalc move = (move, (negate.minimax) (makeMove pos move))
+    moves = map makeMoveCalc (getPossibleMoves pos)
+    ourMove = maximumBy (compare `on` snd) moves
+
+playUntilEnd :: TTTPosition -> (TTTPosition -> Maybe TTTPosition) -> TTTPlayer -> IO ()
+playUntilEnd pos compFunc compStep =
+  case calcMoveState pos of
+    Move player -> if player == compStep then
+      case compFunc pos of
```

Comment:
also could be written as `playUntilEnd (fromMaybe pos (compFunc pos)) compFunc step`